### DISCUSSION TITLE
The cost time of backward between different ranks have big difference with multi-machine training

### DISCUSSION THREAD

**Discussion by wuyujiji** [horovod/horovod#2772] [Q&A]
## Environment:
1. Framework: TensorFlow
2. Framework version: 1.15.0
3. Horovod version: 0.19.2
4. MPI version: 4.0.0
5. CUDA version:10.0
6. NCCL version:2.4.8
7. Python version:3.6.8

## The description of question
Excuse me！I would like to obtain the `forward time`, `backward time (only computing part)`,  `communication time ` and `parameters update time `for each step.  The obtained methods of each stage time are shown as follow:
### Forward timestamp
Adding a `fw_time timestamp op` after the `compute loss op`, as shown in Fig1:

![image](https://user-images.githubusercontent.com/25579435/105513768-ab9bbb00-5d0d-11eb-9233-e92cb01e47ea.png)
### Backward timestamp
Adding a `bw_time timestamp op` after the `tensorflow's compute_gradients op`, as shown in Fig2:

<img width="615" alt="截屏2021-01-22 下午8 02 12" src="https://user-images.githubusercontent.com/25579435/105503164-3aeea180-5d01-11eb-8e0a-0945b6856619.png">

The `tf_grads_and_vars` is obtained by modifying the `compute_gradient` function in `Horovod/tensorflow/__init__.py`. In details, **the result of compute_gradients on the tf optimizer side is also returned**， as shown in Fig3:

<img width="577" alt="截屏2021-01-22 下午8 06 59" src="https://user-images.githubusercontent.com/25579435/105503595-bf412480-5d01-11eb-9bb1-a5761f3b970d.png">

### Communication time (obtain from horovod's timeline)
Directly obtaining the ncclAllreduce time of the first layer's tensor in Horovod's Timeline, as shown in Fig4:

<img width="767" alt="截屏2021-01-22 下午8 13 00" src="https://user-images.githubusercontent.com/25579435/105505423-e567c400-5d03-11eb-895f-435f6258601e.png">

### Parameter update time method
Adding the `bw_time timestamp op` after the `horovod's compute_gradients op`, as shown in Fig5: 

<img width="1047" alt="截屏2021-01-22 下午8 16 16" src="https://user-images.githubusercontent.com/25579435/105505467-f1538600-5d03-11eb-895f-0e87ee67fdf3.png">

Besides the `forward`, `backward` and `parameter update` timestamp, I also record the start timestamp in `before_run` function and end time in `after_run` function. 

<img width="755" alt="截屏2021-01-22 下午8 24 34" src="https://user-images.githubusercontent.com/25579435/105510767-0501eb00-5d0a-11eb-862f-d318bcd3fd84.png">

 Based on these timestamp, I compute the `forward`, `backward`, `update` and `total step` average time of 100 steps by `fw_time - self.t0`, `bw_time - fw_time`, `after_time - update_time` and `after_time - self.t0`. The trained model is BERT Large. However, when printing the log of all ranks in `32 GPUs 4 machines`, I find a strange phenomenon that the backward times of different ranks have big difference, such as `rank11` is `0.488s`, `rank3` is `0.654s`, but the `forward`, `update` and `total step time` is basically the same, as shown in Fig6:

<img width="1443" alt="截屏2021-01-22 下午6 19 17" src="https://user-images.githubusercontent.com/25579435/105517598-367eb480-5d12-11eb-8b2e-dbeabe0b46dd.png">

### 1. the GPU of rank19 has broken or is different from other GPUs?
**No!** In order to verify this point, I run the `single GPU single machine` with rank19 GPU and find the time of each stage is both normal, namely the `forward`, `backward`, `update` and `total step` time is respectively `0.318s`, `0.450s`, `0.088s` and `0.850s`. In addition, I also run the `8 GPUs single machine` and the result looks like the normal, as shown in Fig7:

<img width="792" alt="截屏2021-01-22 下午8 52 01" src="https://user-images.githubusercontent.com/25579435/105510811-13e89d80-5d0a-11eb-9fb7-43a29ca3e76d.png">

### 2. The phenomenon is only occurred in multi-machine training?
**Yes!** For verifying this point, I also did the `16 GPUs 2 machines` experiments and the result is shown as Fig8:

<img width="1169" alt="截屏2021-01-22 下午6 06 58" src="https://user-images.githubusercontent.com/25579435/105511290-b1dc6800-5d0a-11eb-905a-d8c80e8eaf4c.png">

It can be observed that the `backward time` between the `rank2` and `rank7` are respectively `0.490s` and `0.533s`, the `forward` , `parameter update` and `total step time` is basically the same!  In other words, the phenomenon of the `16 GPUs 2 machines` is the same as the `32GPUs 4 machines`.

Furthermore, when I remove the `allreduce_grads` op of `compute_gradients` function, as shown in Fig9:
<img width="561" alt="截屏2021-01-24 上午11 42 08" src="https://user-images.githubusercontent.com/25579435/105620722-c750af80-5e3a-11eb-900d-36ddc70e1dfc.png">

The all rank's results of ` 32 GPUs 4 machines ` are both normal, namely the cost time of each stage basically equals to the` single GPU single machine's`. 

## Ask for help
Based on above experiments, I really can’t analyze the reason. I doubt the reason of the strange phenomenon is caused by horovod？Sincerely ask for help，thanks a lot！！！

**Comment by chongxiaoc** [comment]
Thanks for the detailed analysis.
It looks like the difference is from inter-node nccl allreduce.

I found a similar issue reported in NCCL side:
https://github.com/NVIDIA/nccl/issues/460

@romerojosh Do you expect different ranks in NCCCL end communication at different time separately?

**Comment by wuyujiji** [comment]
@chongxiaoc Thanks for your reply! I also found the phenomenon that the bw_time of inter-node is larger than intra-node according to nccl info channel. But I do not know why the inter-communication can affect the bw_time?

**Comment by chongxiaoc** [comment]
> @chongxiaoc Thanks for your reply! I also found the phenomenon that the bw_time of inter-node is larger than intra-node according to nccl info channel. But I do not know why the inter-communication can affect the bw_time?

horovod requires the allreduce operation to average/sum gradients from all ranks, including intra-node and inter-node. That happens in the last step in `bw` in your benchmark.

https://github.com/horovod/horovod/blob/9727b33542f129fed273902ce727b6b50dfab112/horovod/tensorflow/__init__.py#L475

**Comment by wuyujiji** [comment]
Okay!  The NVIDIA/nccl#460 explains that` if the CPU network thread is competing with other threads and therefore has to wait until the next timeslice (e,g, 200ms), so the final message arrives to a node later in inter-communication`.  Modify by this, I enable the GDR to avoid the CPU network transmission and the `bw_time` of all ranks are both normal. For this phenomenon, I have two points：
1. The reason is indeed as NVIDIA/nccl#460 said that the CPU network thread cannot compete for resources, and the final message arrives to a node later in inter-communication. 
2. The reason may be that the network transmission delay caused the `GPU resource occupation time` of inter-nodes is longer than intra-nodes, and thus no extra GPU resource for backward computation. 

@chongxiaoc Looking forward to your opinion !!

**Comment by chongxiaoc** [comment]
> 1. The reason is indeed as [NVIDIA/nccl#460](https://github.com/NVIDIA/nccl/issues/460) said that the CPU network thread cannot compete for resources, and the final message arrives to a node later in inter-communication.

This sounds more reasonable to me, since you verified that, with the help of GDR to bypass CPU, slow down is gone.

**Comment by wuyujiji** [comment]
Could you please share to analyze why the second point is unreasonable?

**Comment by chongxiaoc** [comment]
I just feel `no extra GPU resource for backward computation` happens really rare. It is not like that GPU is shared by other applications when your training is running.

**Comment by wuyujiji** [comment]
Let me detailed explains the  means of `no extra GPU resource for backward computation`.  When the trained model is  so large (e.g. BERT Large), the GPU Util is consistently 100%, I also proved this in my experiments. However, **when `ncclAllreduce` kernel of each rank is executed，it will also occupy a certain amount of GPU resources**. If the cost time of inter-nodes `ncclAllreduce`  is longer than intra-nodes, it indicates that **the GPU resource occupation time of inter-nodes is longer than intra-nodes**. In this way, the inter-nodes have `no extra GPU resource for backward computation` and the corresponding `bw_time` becomes large.

This is myself opinion, could you have other opinion？

**Comment by chongxiaoc** [comment]
My point is ncclAllreduce and backprop happens in sequential, they are not competing for shared GPU resource on the same rank/machine.

**Comment by wuyujiji** [comment]
https://github.com/NVIDIA/nccl/issues/338 shows that `ncclAllreduce` and `backprop` exist overlap, and  the backprop computes slower when it overlaps `ncclAllreduce` kernel. https://github.com/NVIDIA/nccl/issues/338#issuecomment-698579244 gives his point，so I think it may be the resource problem.

**Comment by chongxiaoc** [comment]
> [NVIDIA/nccl#338](https://github.com/NVIDIA/nccl/issues/338) shows that `ncclAllreduce` and `backprop` exist overlap, and the backprop computes slower when it overlaps `ncclAllreduce` kernel. [NVIDIA/nccl#338 (comment)](https://github.com/NVIDIA/nccl/issues/338#issuecomment-698579244) gives his point，so I think it may be the resource problem.

His example is creating an allreduce outside of training, in order to create the overlapping scenario, to verify the slowdown.
This doesn't happen to your case of using horovod.

Your training script is doing forward->backward->allreduce->optim.step()->forward->... It is sequential steps here. There is no external additional allreduce to compete with your training.

**Comment by wuyujiji** [comment]
For horovod, Is `backward->allreduce`  in sequential? I am not sure I understand correctly. @tgaddair @romerojosh Could you two have spare time to reply? Thanks a lot.

**Comment by tgaddair** [comment]
Hey @wuyujiji, in TensorFlow, the parallelism between the backward pass and the allreduce is controlled by TensorFlow itself via its internal thread pool. If you set TensorFlow's inter-op parallelism to 1, then every op will execute sequentially. But by default, TensorFlow will execute ops that don't depend on one another in parallel. This is the case for most of the backward pass, where we can execute the allreduce calls in Horovod while we are simultaneously computing gradients in the backward pass. This is what allows us to overlap computation and communication.

Even when you execute TensorFlow eagerly, we will construct a subgraph for the allreduce, which allows us to parallelize the computation. You can see what I'm referring to in the code [here](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/__init__.py#L401).

**Comment by wuyujiji** [comment]
> Okay! The [NVIDIA/nccl#460](https://github.com/NVIDIA/nccl/issues/460) explains that` if the CPU network thread is competing with other threads and therefore has to wait until the next timeslice (e,g, 200ms), so the final message arrives to a node later in inter-communication`. Modify by this, I enable the GDR to avoid the CPU network transmission and the `bw_time` of all ranks are both normal. For this phenomenon, I have two points：
> 
> 1. The reason is indeed as [NVIDIA/nccl#460](https://github.com/NVIDIA/nccl/issues/460) said that the CPU network thread cannot compete for resources, and the final message arrives to a node later in inter-communication.
> 2. The reason may be that the network transmission delay caused the `GPU resource occupation time` of inter-nodes is longer than intra-nodes, and thus no extra GPU resource for backward computation.
> 
> @chongxiaoc Looking forward to your opinion !!

@tgaddair Thanks for your reply and I understand it! Besides that, what is your opinion for this problem?

**Comment by tgaddair** [comment]
Do you know if this behavior is specific to the model you're training or your hardware? If you run [tensorflow_synthetic_benchmark.py](https://github.com/horovod/horovod/blob/master/examples/tensorflow/tensorflow_synthetic_benchmark.py), do you get good scaling efficiency up to 32 GPUs?

If the performance on the benchmark is still imbalanced, it may be a hardware / networking issue of some kind. If the performance of the benchmark is good, then we should take a closer look at how the work is distributed in the model.

**Comment by wuyujiji** [comment]
![image](https://user-images.githubusercontent.com/25579435/109592599-7bd0a600-7b4a-11eb-8056-66e1152306e6.png)

I test [tensorflow_synthetic_benchmark.py](https://github.com/horovod/horovod/blob/master/examples/tensorflow/tensorflow_synthetic_benchmark.py), it also has this problem.  The reason may be the ncclAllreduce kernel, detailed in my submitted issue:   https://github.com/NVIDIA/nccl/issues/470
